Extract a ColorPanel component as a reusable component between Global Styles and Block Inspector

youknowriad opened the pull request:
Related to #37064 and #47348 and similar to #47356, #48636 and #48070

## What?

A year ago now, I've opened a discussion in #37064 that proposes unifying the components that are used in the block inspector and global styles. The idea is that we'd have components that look like this:

```
function SomePanel( { value, onChange, inheritedValue } ) {

}
```

That component would render a panel used to edit a given style (say "typography" or "colors"), the `value` has a well specified prop (that is mapped over the "style" object of a block or a global style "node" ). The idea is that we could use that exact same panel in both places, the only difference is a wrapper that map the "value" to the right place:
 - block attributes for the block inspector
 - theme.json (global styles) for the global styles UI

Recently we realized this vision for the TypographyPanel, BorderPanel and DimensionsPanel and the current PR does the exact same thing for the ColorPanel and uses the new UI component for both inspector controls and global styles.

ColorPanel is the last panel to migrate but also probably the most complex one, unfortunately this PR turns out to be a bit big, and it's unclear whether we can split it into smaller PRs.

Since the goal of this PR is also to bring parity with Global Styles, It ended up including a new feature: You can now set "hover" colors for links within the block inspector.

**To do**

 - [x] Figure out what to do with "heading" and "button" element colors and whether we want to add support for these to block "local" styles as well as "global styles".
 - [x] Cleanup the styles of the components: Right now I've reused some classnames from other components. Ideally we'd find a way to use `ItemGroup` and `Item` and find a way to compose these components within `PanelTools` and `PanelToolsItem`. This would allow us to remove a bunch of CSS and classnames but I haven't found a decent way to do that. Maybe someone familiar with these PanelTools component could help here. 
 - [x] Test properly that there's no unexpected differences in behavior between this PR and trunk. (especially regarding the preset colors...)

## Testing Instructions

 - you can test the "color" panels in global styles and ensure that they still work as expected. 
 - you can also test the "color" panels in the block inspector.

youknowriad committed:
Color panel reusable panel

youknowriad committed:
Add link hover color support to blocks

youknowriad committed:
Use the new ColorPanel in Global Styles too

youknowriad commented on lib/block-supports/elements.php:
```diff
@@ -94,6 +99,16 @@ function gutenberg_render_elements_support_styles( $pre_render, $block ) {
+			)
+		);
+	}
+
```
I was surprised that I had to make changes to this file to make the "hover" colors styles work in the frontend. I thought they'd just work since we support them for Global Styles. I guess I'll have to do similar changes for heading and button elements which makes me think that there's probably a better way to avoid all this logic and rely on some existing code (style engine or whatever)

andrewserong replied:
The style engine is currently naive about the selector to use, so it takes a style object and passes back styles using the selector it's given. This created a fairly clean separation between block supports and the style engine, while containing scope for the style engine project.

Perhaps, a bit like the typography block support, we could have a function in `elements.php` that is responsible for dealing with each of the potential element selectors, and then calls the style engine. At the moment that's this function, but it could probably be extended to deal with potential future elements. And we could then use that function in both global styles and at the block level.

For the moment, though, inlining an additional call to the style engine like this sounds good to me!

youknowriad replied:
Maybe the style engine had two methods. The current naive one and what that takes a full "style" object and generate everything including selectors.

andrewserong replied:
> that takes a full "style" object and generate everything including selectors.

Good point — it'd be worth exploring!

youknowriad commented on packages/block-editor/src/hooks/border.js:
```diff
@@ -163,7 +163,6 @@ export function BorderPanel( props ) {
 		<StylesBorderPanel
 			as={ BordersInspectorControl }
 			panelId={ clientId }
-			name={ name }
```
These were just leftovers.

youknowriad commented on packages/block-editor/src/hooks/contrast-checker.js:
```diff
@@ -6,45 +6,20 @@ import { useState, useEffect } from '@wordpress/element';
 /**
  * Internal dependencies
  */
-import ColorGradientSettingsDropdown from '../components/colors-gradients/dropdown';
```
Unfortunately, I can't remove the old components as they're exposed and used in several core and third party blocks.

youknowriad commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -384,39 +384,40 @@ const withElementsStyles = createHigherOrderComponent(
+						? props.attributes.style?.elements?.link?.[ ':hover' ]
+						: undefined,
+					selector: `.editor-styles-wrapper .${ blockElementsContainerIdentifier } ${ ELEMENTS.link }:hover`,
+				},
```
Same as https://github.com/WordPress/gutenberg/pull/48893/files#r1128026980 Feels like this should be somehow part of style engine or whatever.

youknowriad committed:
Add elements support to the color panel

youknowriad commented:
I've added the "headings" and "buttons" colors to the color panel component. I noticed a problem though, for a reason that I can't explain, the "heading" and "button" buttons are not showing up. If I remove them from the "DEFAULT_CONTROLS", I can start adding them from the more menu but if they're made default controls, they are not shown at all. I think it's a bug in `ToolsPanel` component but I really don't understand it. Maybe @aaronrobertshaw or @ciampo would know.

aaronrobertshaw committed:
Fix display of button and heading elements

aaronrobertshaw committed:
Restore layout of global styles color panel items

aaronrobertshaw commented:
Thanks for tackling the colors panel @youknowriad, it's always been rather convoluted. When this PR is ready for final reviews it might help to have a bit of a breakdown around the new approach, what changes were made and why etc.

> I noticed a problem though, for a reason that I can't explain, the "heading" and "button" buttons are not showing up

The `hasElement` function passed via the `hasValue` prop doesn't return a `boolean` as the `hasValue` prop requires. Tweaking that, makes the button and heading elements show up for me. I haven't had a chance to do any thorough testing yet but at first blush, it appears to work ok.

One possibility to guard against this might be to update the ToolsPanelItem to check for the item's property rather than the property `!== undefined`.

In addition, there were a couple of missing styles and props for the global styles panel.

I've taken the liberty to push a couple of commits to address these. Please feel free to revert or change anything you'd like to improve.

I can give this PR a closer look later this evening or early next week. 

Nice work so far!


https://user-images.githubusercontent.com/60436221/224200104-42a17d7f-2227-4aa3-a3d4-4fbfe26a91c9.mp4

youknowriad committed:
Add captions

youknowriad committed:
Fix styles

youknowriad commented:
I've rebased the PR and added the new caption element. This is getting very close and I'd appreciate some reviews here. Thank you ❤️

jasmussen commented:
Nice work. Taking this for a spin just to compare with trunk, I'm focusing mostly on Global Styles → Colors. Text appears the same. Buttons appears the same. Headings appears the same. 

<img width="274" alt="headings after" src="https://user-images.githubusercontent.com/1204802/226600342-696e2505-4724-40d0-95f2-919493542475.png">

It's coming into clarity that the text/heading hierarchy on this page is not that considered. The tabs for heading background are also awkwardly indented. So there are some larger design tweaks we can do, separately.

I think I'm missing something, though, because for links I see hover both in this PR and in trunk:

<img width="469" alt="links after" src="https://user-images.githubusercontent.com/1204802/226601064-072ac023-c743-41a0-885d-d2e4540f4a8e.png">

I'm also seeing "Elements" in the main GS panel, instead of "Color" as Aaron appears to see. Did I test the PR right?

<img width="270" alt="Screenshot 2023-03-21 at 13 06 04" src="https://user-images.githubusercontent.com/1204802/226601181-778c6064-7ef4-446e-b518-33ba7412ab9e.png">

youknowriad commented:
@jasmussen yeah I think there's something missing in your tests, are you sure everything built properly and all?

jasmussen commented:
Oh, I found it. In some testing I had deactivated the gb plugin. My apologies, as you were. 

Here it's working:

![status](https://user-images.githubusercontent.com/1204802/226603944-281d51bc-2a21-459b-8aee-57cdcfaea809.gif)

Generally looking good. It's okay that these don't also have the palette picker drilldown, so long as the Palette option already exists. In fact on a separate note, we should only have the palette visible in GS > Colors, not in GS > Blocks > [Blockname] > Color:

<img width="336" alt="Screenshot 2023-03-21 at 13 20 39" src="https://user-images.githubusercontent.com/1204802/226604387-6f889eea-3801-4f04-9e3b-76587eb02958.png">

Two small glitches. Visible in the GIF, occasionally I could get the flyout to open in an awkward place. It seemed to be when clicking on or about the swatch inside the itemgroup:

<img width="574" alt="Screenshot 2023-03-21 at 13 18 57" src="https://user-images.githubusercontent.com/1204802/226604542-1c6f4587-0a79-48d7-92ef-fe24e3c76286.png">

Not sure if that's from this branch or general.

Otherwise, this separator:
<img width="290" alt="Screenshot 2023-03-21 at 13 19 05" src="https://user-images.githubusercontent.com/1204802/226604587-01f48712-02a7-487e-96fd-4096a99ecdbe.png">

Either it should go edge to edge left and right, or it should just be removed. Both would be fine.

andrewserong commented:
> Either it should go edge to edge left and right, or it should just be removed. Both would be fine.

I noticed the separator, too. Looks like it might be unintentional for it to be there in global styles?

andrewserong reviewed (commented):
Great work here so far! Aside from what Joen mentioned above, it's testing quite well for me. It took me a little while to find the `h1` through `h6` controls for adjusting colors for each of the different headings, but once I did, I really liked how compact it looks seeing them all lined up in the one list 👍

<img width="897" alt="image" src="https://user-images.githubusercontent.com/14988353/226802724-6c4f6460-c772-4dd4-911a-f55582203ab6.png">

---

Just looking into some of the e2e failures, and it looks like the style variations test failures are because of losing the aria-label on the background color button elements.

Here's the visual difference between the two panels in trunk vs this PR

| Trunk | This PR |
| --- | --- |
| <img width="274" alt="image" src="https://user-images.githubusercontent.com/14988353/226801286-3e9c282d-23a7-46cd-bb3b-35091d3027a3.png"> | <img width="276" alt="image" src="https://user-images.githubusercontent.com/14988353/226801260-2836adfd-3e50-4644-a0e5-661dc5d1ab47.png"> |

And here's the difference in markup for the buttons

Trunk:

<img width="755" alt="image" src="https://user-images.githubusercontent.com/14988353/226801381-ff8365c4-dc2f-434d-a644-f57886a1c5dd.png">

This PR (missing aria-label):

<img width="784" alt="image" src="https://user-images.githubusercontent.com/14988353/226801424-5a0d1cc9-dc50-4c8a-9830-2d95c0bc752d.png">

The test is [currently looking](https://github.com/WordPress/gutenberg/blob/97cf79af9144d5aa585fc7f2f07ec736251f2e3f/test/e2e/specs/site-editor/style-variations.spec.js#L130) for the text `Colors background styles`, but perhaps this could be updated to `Background` since there's no `aria-label` now? Or should we look at restoring the `aria-label`?

youknowriad committed:
Remove border and reorder elements

youknowriad committed:
Fix e2e test

youknowriad commented:
I removed the border but I wasn't able to reproduce the popover position issue cc @jasmussen 

It would also be good to test in the block inspector if there's any unwanted impact.

andrewserong commented on packages/block-editor/src/components/global-styles/hooks.js:
```diff
@@ -42,6 +41,11 @@ const VALID_SETTINGS = [
 	'color.link',
 	'color.palette',
 	'color.text',
+	'color.gradient',
```
Is this line intended? We already have `color.gradients` further up, so I wasn't sure if this was a typo.

aaronrobertshaw replied:
Yeah, I wasn't sure if it was a typo or a new setting given some of the odd requirements to disable color UIs. In terms of the [schema](https://github.com/WordPress/gutenberg/blob/77bb6c8f541d9c5097cf4865e478e04f69b20f7b/schemas/json/theme.json#L179) at least, it's not a valid setting. 

There are several other follow-on places that use `color.gradient` in this PR that would need tweaking if this is a typo.

The abundance of gradient settings might warrant clarification in our docs and the schema. We already have `gradients` and `customGradient` under settings. The block support flags use `gradients` as well, I believe, although there is a typo in the [block support docs](https://github.com/WordPress/gutenberg/blob/0c208a7a1f730b546ecb9f38bc7b22e1c892c3a4/docs/reference-guides/block-api/block-supports.md#colorgradients) that reference color.gradient.

youknowriad replied:
This is not a typo. It's been some time that I did this change but the idea is that `color.gradients` is the equivalent of `color.palette` for solids and that `color.gradient` is the equivalent of `color.text` or `color.background` (enable/disable these things)

youknowriad replied:
Now, that I think about it. I'm not sure why I had to add a new key instead of relying on `color.background` and the gradients preset. I'll have to check again.

youknowriad replied:
Ok, I've checked this and I remove the setting for now. That said, I think we should restore it at some point. Some blocks can today disable "gradients" but enable "backgroundColor". The opposite is not possible, basically there's an inconsistency with the settings. Anyway, a thing for the future.

andrewserong reviewed (commented):
I re-tested in the block editor and ran into a couple of subtle / minor issues.

The contrast checker appears to be behaving inconsistently for me. Once it's displayed, it doesn't appear to update until I click away from the block and then back into the block again:

https://user-images.githubusercontent.com/14988353/227104923-70e477be-b654-4640-9321-8f4b5bf034fc.mp4

---

In its placeholder state, before any colors have been selected, on `trunk` we display the cross through the color indicator. With this PR applied, it looks as if white has been chosen for the colors:

| Trunk | This PR |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/227105126-12820097-c72b-436b-8489-7acce668a0a9.png) | ![image](https://user-images.githubusercontent.com/14988353/227105164-ced8f028-39c3-4b67-a7c2-7bcbea5b79f0.png) |

(As an aside, cool to see that this fixes up the issue where on `trunk` the list of experimental default controls isn't respected, whereas here `link` is hidden, consistent with the paragraph block's `block.json`).

---

I was also able to reproduce the odd Popover position issue that Joen mentioned. Unfortunately, I haven't been able to work out _why_ it's doing it, and it appears inconsistent, but here's a video of clicking in and out of the Link color, where it seems to randomly display correctly and at other times too far to the right:

https://user-images.githubusercontent.com/14988353/227105513-abbbf1f1-9649-4aab-865b-91631110c597.mp4

aaronrobertshaw commented on lib/theme.json:
```diff
@@ -11,6 +11,11 @@
 		},
 		"color": {
 			"background": true,
+			"gradient": true,
```
Is this a new setting? It's not in the theme.json schema and we already have `customGradient` or `gradients`.

aaronrobertshaw reviewed (commented):
I've taken this for another spin and hit an issue in addition to what Andrew noted in his review.

The issues I encountered were:
- Elements aren't resetting via panel menu's reset all item
    - Occurs in both Global Styles & Block Inspector
    - Individual menu items correctly reset the control & value
    - Includes link in Block Inspector, and caption, button, heading, h1-h6 in Global Styles
- Contrast checker not displaying upon color selection
- Missing diagonal slash through empty color indicator
- Incorrectly positioned color popovers (really random)

| Block Inspector | Global Styles | Popover Positioning |
|---|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/227119716-34046533-6560-43d8-af7f-9a923749ccdd.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/227119885-3d5edb5e-bfc3-45b4-8349-75e678de490c.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/227121200-ae996efc-77ff-4ae6-8645-d62d51e65a7f.mp4" /> |

youknowriad committed:
Fix undefined color indicator

youknowriad committed:
Remove dead CSS code

youknowriad committed:
Fix color contrast checker hook dependencies

youknowriad committed:
Fix resetAll

youknowriad committed:
Remove gradient key

youknowriad commented:
Ok. I've addressed all the issues I think aside the popover one.

The Popover issue is very very hard to trigger for me (I did manage to do it a couple times) but it's also present on trunk for the block inspector colors, so it's not something introduced in this PR.

youknowriad committed:
Remove more dead code

youknowriad committed:
Remove dead code

MaggieCabrera commented on packages/edit-site/src/components/global-styles/screen-colors.js:
```diff
@@ -263,40 +52,12 @@ function ScreenColors( { name, variation = '' } ) {
 				<VStack spacing={ 10 }>
 					<Palette name={ name } />
```
I'm not sure if it's out of the scope of this PR, but I found [this issue](https://github.com/WordPress/gutenberg/issues/36541#issuecomment-1480882023). Is it easier with this refactor to know if we are on the global vs the block level to hide the palette component if I'm on the block level?

youknowriad replied:
Yeah, it's a bit independent. The current PR doesn't touch the "palette" global styles screens. It only refactors the UI elements used to actually set the styles.

ciampo commented:
Re the popover bug, from the video it seems that it happens only when the popover contains a tabbed interface (e.g "Solid" and "Gradient" tabs for the background color, or "Default" and "Hover" tabs for the "Link" color). Maybe it can help to narrow down the issue.

aaronrobertshaw reviewed (commented):
Thanks for the continued effort here @youknowriad 👍 

> Ok. I've addressed all the issues I think aside the popover one.

I've given this another test. 

All but one of the issues I raised in my last review has been resolved.

❓ Reset all now clears heading, button, caption etc styles. It doesn't however reset link hover colours correctly in either the block inspector or global styles.
✅ Contrast checker now displays when making a poor contrast selection and clears appropriately.
✅ The diagonal slash through empty color indicator is now shown correctly
✅ Incorrectly positioned color popover is appearing on trunk for me and not introduced in this PR.

| Block Inspector | Global Styles |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/227397326-9feb001c-4a95-409d-9802-eab4060236fb.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/227397311-c538f09a-9326-4a06-b691-d496b0d4be3f.mp4" /> |

youknowriad commented:
> ❓ Reset all now clears heading, button, caption etc styles. It doesn't however reset link hover colours correctly in either the block inspector or global styles.

Oh yeah, forgot about that one, thanks. It should be a quick fix.

youknowriad committed:
Reset hover colors properly

youknowriad commented:
The link hover reset is fixed now.

ciampo commented:
Opened https://github.com/WordPress/gutenberg/issues/49349 to track the Popover glitch

andrewserong commented on lib/theme.json:
```diff
@@ -11,6 +11,10 @@
 		},
 		"color": {
 			"background": true,
+			"heading": true,
+			"button": true,
+			"text": true,
+			"caption": true,
```
Tiniest of nits: prior to this change the keys of this object were in alphabetic order. Did we want to preserve that?

youknowriad replied:
Eagle eye, I guess there's argument for grouping similar keys but let's keep the alphanumeric order for now.

andrewserong commented on packages/block-editor/src/components/global-styles/hooks.js:
```diff
@@ -42,6 +41,10 @@ const VALID_SETTINGS = [
+	'color.background',
+	'color.heading',
+	'color.button',
+	'color.caption',
```
Tiny nit: the `color.` settings were in alphabetical order prior to this change.

andrewserong commented on packages/block-editor/src/components/global-styles/color-panel.js:
```diff
@@ -0,0 +1,754 @@
+			} );
+		};
+		const supportsTextColor = true;
+		const supportsBackground = name !== 'caption';
```
Is it worth adding a comment about why this is a one-off for `caption`? I could imagine this being tricky to find, or potentially confusing for someone else looking at this. From memory, background color support for `caption` is disabled because it doesn't have padding support?

```suggestion
		// Background color is not supported for `caption` as there isn't yet a way to set padding for the element.
		const supportsBackground = name !== 'caption';
```

andrewserong commented on packages/block-editor/src/components/global-styles/color-panel.js:
```diff
@@ -0,0 +1,754 @@
+				supportsTextColor && supportsBackground
+					? [
+							elementTextColor,
+							elementBackgroundColor ?? elementGradient,
```
Should this line be the other way around? This preferences background color to be displayed before gradient color. So, in TwentyTwentyThree theme, which has a background color of white set in its `theme.json` file, if I set a gradient color, then it gets displayed here as white:

<img width="270" src="https://user-images.githubusercontent.com/14988353/227818397-ab6034bc-6a3b-4b10-adbb-e1920184fb05.png" />

If I remove the background color from the `theme.json` file, then it displays the gradient as expected:

<img width="270" alt="image" src="https://user-images.githubusercontent.com/14988353/227818459-4565e060-353b-4021-a325-018af1b4c577.png">

andrewserong commented on packages/block-editor/src/components/global-styles/color-panel.js:
```diff
@@ -0,0 +1,754 @@
+					: [
+							supportsTextColor
+								? elementTextColor
+								: elementBackgroundColor ?? elementGradient,
```
Same as above, should these be flipped so that the gradient is preferenced before the background color?

andrewserong reviewed (approved):
Just took this for another spin, and it's looking in quite good shape! I only found one real issue, which looks like it should be a fairly simple fix 🤞

✅ Disable `settings.color.heading` works correctly
✅ Disable `settings.color.button` works correctly
✅ Disable `settings.color.caption` works correctly
✅ ContrastChecker now behaves as on `trunk` (it's a little flaky, but doesn't appear to be any flakier than on trunk)
✅ Setting / resetting individual color controls / reset all at the block level appears to work
✅ Setting / resetting individual color controls / reset all at the elements level in global styles appears to be working
✅ Setting / resetting individual color controls / reset all at the block level in global styles appears to be working
❓ The only real issue I encountered was with the color indicator not displaying gradient color for Background when set. This appears to be because TwentyTwentyThree theme sets a background color, so when a gradient is set, both values exist, and the background color is preferenced in this PR for the indicator over the gradient color. On trunk, I think it's the other way around.

Other than that, after reading through the code again, and with the known issue with the popover, I think this PR's looking in good shape! It's a challenging one to review given the size of the change, so it'd probably be good to get another approving review before landing it, just in case I've missed anything.

Also, there's probably an opportunity to consolidate or split out some of the logic in the `ColorPanel` component in follow-ups, as it's quite a long component. My main note there is that it might be good to see if we can improve the readability to make it easier to make subsequent changes.

For now, though, great work on all the consolidation efforts here, and for the persistence across each of these components and panels! ✨

kevin940726 commented on test/e2e/specs/site-editor/style-variations.spec.js:
```diff
@@ -86,13 +86,13 @@ test.describe( 'Global styles variations', () => {
 		await expect(
 			page.locator(
-				'role=button[name="Colors background styles"i] >> data-testid=background-color-indicator'
+				'role=button[name="Background"i] >> .component-color-indicator'
```
Why do we have to use class name selectors here? Could we use role selectors or `data-testid` instead?

youknowriad replied:
I was expecting this comment :) 

I preferred classnames for two reasons here:

 - First one, I don't like data-test-ids attributes when they linger on production code. I think we should avoid changes in production code that are only meant for testing. One potential solution would be to strip these in non test builds but not sure how to do that.
 - Second one is that the change in the structure of the component (more generic) means we can't specifically say that a particular indicator is "background" or other "text" unless we add extra props and extra complexity that I was not happy to add for test purposes.

kevin940726 replied:
Let's not start a conversation here about whether `data-testid` is a best practice in tests. We already have [best practices](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/e2e/README.md#use-accessible-selectors) guidelines committed to the repo, as well as all the tooling recommendations in the community ([testing-library](https://kentcdodds.com/blog/making-your-ui-tests-resilient-to-change#whats-with-the-data-testid-query), [playwright](https://playwright.dev/docs/locators#locate-by-test-id), etc), so I would expect us to follow them here :). You are more than welcome to open an issue or discussion and we can discuss further and rethink our best practices there!

> Second one is that the change in the structure of the component (more generic) means we can't specifically say that a particular indicator is "background" or other "text" unless we add extra props and extra complexity that I was not happy to add for test purposes.

I'm afraid that I don't follow, could you provide me an example code of what you mean here?

youknowriad replied:
We pass an array of colors to the indicators component and we don't specificy which color is which. So there's no way I can say which indicator is background or text here.

youknowriad replied:
I wanted to open a PR to update the data test id best practice and I don't see it mentioned anywhere. Did I miss anything? Also, it seems it's the only place where it was used. So I guess now there's nothing to change for this?

talldan replied:
If I'm testing the right thing, then I think there are deeper accessibility issues with this interface.

The button doesn't have any accessible label to say what color its set to, so that's maybe one of the reasons the test has to do something weird.

I tried testing in voiceover/safari to see what it's like, and it seems like you can't actually move focus into the color picker popover:

https://user-images.githubusercontent.com/677833/227924738-58352a14-5bd4-45a3-8968-53a37d0ec676.mp4

youknowriad replied:
This UI is the same UI we already had in the block inspector so I don't expect the UI to be completely inaccessible, that said a small bug is possible. I can reproduce the fact that the focus doesn't seem to jump within the popover when clicked. That said the label is there because buttons use the text content as label when there's no aria-label and I believe the text in these buttons is good enough to avoid aria-labels.

I'll look at the focus issue.

youknowriad replied:
That was a bug in the TabPanel component that should be fixed with this PR https://github.com/WordPress/gutenberg/pull/49368

talldan replied:
Thanks, appreciate the quick fix. I've merged that PR.

> That said the label is there because buttons use the text content as label when there's no aria-label and I believe the text in these buttons is good enough to avoid aria-labels.

Accessibility is at least a bit better now that the popover content can be focused to check what the current color is.

It doesn't necessarily need an aria-label, I was thinking more along the lines of some extra hidden text (`<VisuallyHidden>(current color: #FFF)<VisuallyHidden>`). At the moment it'd be a bit of a frustrating process to determine which colors are used for each element if a user can't see them.

youknowriad replied:
@talldan Yeah, I think that could be a nice addition, but the only labelled item is the whole button, so we need to generate a long description, something like:

 ```
 <VisuallyHidden>
 text color is #FFF, background color is #000, hover text color is #333 and background hover color is #666
 </VisuallyHidden>
```

It needs to also understand gradients. So not a small addition.

talldan replied:
Good point about gradients 😬

youknowriad committed:
Fix gradient indicator

youknowriad commented:
>  The only real issue I encountered was with the color indicator not displaying gradient color for Background when set. This appears to be because TwentyTwentyThree theme sets a background color, so when a gradient is set, both values exist, and the background color is preferenced in this PR for the indicator over the gradient color. On trunk, I think it's the other way around.

Right! I actually was able to reproduce this issue on trunk too, it's fixed now on this branch.

youknowriad committed:
alphanum order in theme.json file

youknowriad committed:
Fixes per review

youknowriad commented:
Thanks for the reviews folks, I'm going to merge this PR as I plan to work on small changes on top of it. 

I'm happy to adapt in follow-up PRs if needed. I really appreciated the reviews as always.

tellthemachines commented on packages/block-editor/src/hooks/color.js:
```diff
@@ -273,194 +218,112 @@ export function addEditProps( settings ) {
+		style: cleanEmptyObject( updatedStyle ),
+		textColor: textColorSlug,
+		backgroundColor: backgroundColorSlug,
+		gradient: gradientSlug,
```
Are we purposely setting all the attributes at once in order to clean out any non-applicable ones, even if the block doesn't support those attributes? Asking because the question arose in #53941.

youknowriad replied:
I think that's the case but to be honest, it's an old PR so no precise memories at the moment.
